Replace deprecated operator overloads in Map and container modules

don-clugston-sociomantic opened the pull request:
In dmd 2.088, the D1 operator overloads are deprecated. This causes almost 2000 deprecation warnings when compiling most apps using our libraries.

This PR deals with the straightforward cases, where the operator is a member of a struct. In this case it is a simple substitution, that won't break any existing code.

(The cases where the operator is a member of a class are far more difficult, because the D1 operator overloads are virtual, whereas the D2 overloads are static. Those cases are not dealt with in this PR).

Additionally in this PR I have fixed one 'shadowing' warning in a unit test, and cleaned up a broken doc comment.

don-clugston-sociomantic committed:
Add unittest for container.more.Stack

don-clugston-sociomantic committed:
Replace deprecated opXXXAssign in Stack

Replace them with opOpAssign.

don-clugston-sociomantic committed:
Replace opCatAsssign with opOpAssign!"~" in Vector

don-clugston-sociomantic committed:
Replace opCatAssign with opOpAssign!"~" in VoidBufferAsArrayOf

Since opCatAssign is deprecated.

don-clugston-sociomantic committed:
Remove typo junk from doc comment

don-clugston-sociomantic committed:
Replace opCatAssign -> opOpAssign!"~" in Heap

Since opCatAssign is deprecated.

don-clugston-sociomantic committed:
Rename variable in unittest to avoid shadowing

The opApply 1 test created a variable 'i' which was then shadowed
in the opApply 2 test. Rename the first one.

don-clugston-sociomantic committed:
Replace opIn_r -> opBinaryRight in Map modules

opIn_r is deprecated. These are structs so the fact that the functions
are non-virtual is not a problem.

don-clugston-sociomantic committed:
Replace opIn_r -> opBinaryRight!"in" in HttpVersionIds

Since opIn_r is deprecated. HttpVersionIds is a struct so there is
nop problem with the fact that it is non-virtual.

don-clugston-sociomantic committed:
Replace opCatAssign -> opOpAssign!"~" in DateTime_tango

opCatAssign is deprecated. This is a private struct so the replacement
is straightforward.

alaksiej-stankievicx-dunnhumby commented on src/ocean/util/container/more/Vector.d:
```diff
@@ -32,7 +32,7 @@ struct Vector (V, int Size = 0)
         alias add       push;
         alias slice     opSlice;
-        alias push      opCatAssign;
+        public template opOpAssign(string op : "~") { alias opOpAssign = push; }
```
Don't we prefer a different way:
```d
public template opOpAssign (string op) if(op == "~") {alias opOpAssign = push;}
```

don-clugston-sociomantic replied:
What we want here is not a constrained template but a specifically defined one (all parameters specified).
The compilation time is substantially better with specialisation, the code is shorter and more readable, and the behaviour is the same.

alaksiej-stankievicx-dunnhumby reviewed (commented):
So, IIRC,  we prefer template constraints not template specification in D2. Do you have any reason to do this in the current way?

stefan-koch-sociomantic commented:
template constraints get evaluated by the CTFE whereas specialization gets resolved in the template dedicated code.
It leads to compile time savings not to use constraints.

alaksiej-stankievicx-dunnhumby commented:
I'm mostly about that in other PR for `ocean` we have constraints (e.g see this https://github.com/Geod24/ocean/blob/6f2181843bbd033f534489be78f3d055391206c7/src/ocean/core/BitArray.d#L591). And if there is no heavy drawbacks I prefer consistent approach.

stefan-koch-sociomantic commented:
it also looks nicer.

don-clugston-sociomantic commented:
Good point. I don't think there's any reason those ones were done in that way.

Geod24 commented:
> Good point. I don't think there's any reason those ones were done in that way.

AFAIR I tried specialization first (it looks nicer) but AFAIR it didn't work. I can't quite remember the specific though.

don-clugston-sociomantic committed:
Use template specialization for operator in BitArray

Instead of using constraints, for the case where only a single operation
is being defined.

don-clugston-sociomantic committed:
Use specialization for "~" operator in Buffer modules

Instead of template constraints.

don-clugston-sociomantic commented:
Repushed with the other concatenation operators changed to use specialization. If it doesn't work, hopefully we'll see why.

don-clugston-sociomantic committed:
Change opAdd/sub to opOpAssign!"+" in IRegisteredClients

Even though IRegisteredClients is a class, these functions were marked
as `final` so the fact that the D2 overloads are non-virtual is
not a problem.

don-clugston-sociomantic committed:
Change opSub -> opBinary!"-" in ProcVFS

Since opSub is deprecated in D2.089.
Time is a struct so the fact that the new overload is not virtual
is not a problem.

don-clugston-sociomantic commented:
I found two more modules with overloaded operators that can be changed to D2-style without problems.

ben-palmer-sociomantic commented:
Changes LGTM. I haven't checked what the unit test coverage is for the changes though. Could be argued to be out of scope of the PR but would be nice to check.

don-clugston-sociomantic commented:
The code coverage actually increases as a result of this PR (since a new unittest is added). But the *reported* code coverage goes down, because we're replacing ordinary functions with templated functions.
The code coverage stats ignore templates. So you get no credit for testing them. 
It's a structural bug in code coverage.

ben-palmer-sociomantic commented:
Any other comments before I merge?
